### PR TITLE
[RLlib] Properly serialize and restore StateBufferConnector states for policy stashing

### PR DESCRIPTION
**Pull request by gjoliver**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/connectors/agent/state_buffer.py [hunk 1]
```diff
@@ -70,11 +89,18 @@ def transform(self, ac_data: AgentConnectorDataType) -> AgentConnectorDataType:
         return ac_data
 
     def to_state(self):
-        return StateBufferConnector.__name__, None
+        # Note(jungong) : it is ok to use cloudpickle here for stats because:
```
Comment (kouroshHakha) [inline rllib/connectors/agent/state_buffer.py:92]: The reason to use cloudpickle over pickle is simply that you are pickling a data-structure that contains lambda function, right?

File: rllib/connectors/agent/state_buffer.py [hunk 2]
```diff
@@ -8,21 +10,38 @@
...
         self._states = defaultdict(lambda: defaultdict(lambda: (None, None, None)))
+        if states:
+            try:
+                self._states = cloudpickle.loads(states)
+            except pickle.UnpicklingError:
+                # StateBufferConnector states are only needed for rare cases
+                # like stashing then restoring a policy during the rollout of
+                # a single episode.
+                # It is ok to ignore the error for most of the cases here.
+                logger.info(
```
Comment (kouroshHakha) [inline rllib/connectors/agent/state_buffer.py:41]: I don't get why you should not error out all the time here? When will you ever pass in a states object in that is ok if it's not unpickled?

**Review by kouroshHakha — APPROVED** [review]
The PR seems to solve the problem with 100-policy test at least. I wonder if it still works with the old numbers. @gjoliver Can you confirm that?

File: rllib/connectors/agent/state_buffer.py (hunk 1 above)
Comment (gjoliver) [reply rllib/connectors/agent/state_buffer.py:92]: I actually think we should always use cloudpickle.
It's not officially guaranteed, but there is a less chance of not being able to restore states saved with a higher version of python if we use cloudpickle.
cloudpickle kinda does it automatically for you, using the right pickle library if the version is low (pickle5 etc).

File: rllib/connectors/agent/state_buffer.py (hunk 2 above)
Comment (gjoliver) [reply rllib/connectors/agent/state_buffer.py:41]: when you recover a policy for serving, we wouldn't need the state buffer state.
actually this comment reminded me, I am letting StateBufferConnector to clear state whenever someone switch it into eval mode.

**Comment by gjoliver** [comment]
> The PR seems to solve the problem with 100-policy test at least. I wonder if it still works with the old numbers. @gjoliver Can you confirm that?

yes, it works with original numbers. just very slow right now because we are doing a lot of unnecessary policy stashing.
I will file a separate issue for that.

File: rllib/connectors/agent/state_buffer.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/connectors/agent/state_buffer.py:41]: In that case we should not pass in the states at all?

Comment (gjoliver) [reply rllib/connectors/agent/state_buffer.py:43]: maybe we can discuss a bit.
I feel like the only case we need to restore this is when policies are stashed and recovered in the middle of an episode.
we don't need this state even for training as long as we keep the policy in cache throughout an episode.
so maybe a better fix is to ping any "in-use" poclies, and prevent them from being stashed.

**Comment by gjoliver** [comment]
tests look good finally with and without the flag flip.
gonna merge now.
thanks.

**Pull request merged**